src/sema/expression/function_call.rs

```diff
@@ -2620,3 +2546,75 @@ pub(crate) fn function_type(func: &Function, external: bool, resolve_to: Resolve
         }
     }
 }
+
+/// This function evaluates the arguments of a function call with either positional arguments or
+/// named arguments.
+fn evaluate_argument(
+    arg: &pt::Expression,
+    context: &ExprContext,
+    ns: &mut Namespace,
+    symtable: &mut Symtable,
+    arg_ty: &Type,
+    errors: &mut Diagnostics,
+    cast_args: &mut Vec<Expression>,
+) -> bool {
+    expression(arg, context, ns, symtable, errors, ResolveTo::Type(arg_ty))
+        .and_then(|arg| arg.cast(&arg.loc(), arg_ty, true, ns, errors))
+        .map(|expr| cast_args.push(expr))
+        .is_ok()
+}
+
+/// This function finishes resolving internal function calls. It returns None if it is not
+/// possible to resolve the function.
+fn resolve_internal_call(
+    loc: &Loc,
+    function_no: usize,
+    context: &ExprContext,
+    resolve_to: ResolveTo,
+    virtual_call: bool,
+    cast_args: Vec<Expression>,
+    ns: &Namespace,
+    errors: &mut Diagnostics,
+) -> Option<Expression> {
+    let func = &ns.functions[function_no];
+
+    if func.contract_no != context.contract_no && func.is_private() {
+        errors.push(Diagnostic::error_with_note(
+            *loc,
+            format!("cannot call private {}", func.ty),
+            func.loc,
+            format!("declaration of {} '{}'", func.ty, func.name),
+        ));
+
+        return None;
+    } else if let (Some(base_no), Some(derived_no)) = (func.contract_no, context.contract_no) {
+        if is_base(base_no, derived_no, ns) && matches!(func.visibility, Visibility::External(_)) {
+            errors.push(Diagnostic::error_with_note(
+                *loc,
+                "external functions can only be invoked outside the contract".to_string(),
+                func.loc,
+                "function defined here".to_string(),
```

Comment:
"here" is not great, also it might not be a definition (e.g. interface). How about reusing the note from above:

`format!("declaration of {} '{}'", func.ty, func.name),`